pydantic/_analyzed_type.py

```diff
@@ -0,0 +1,217 @@
+        ...
+
+    @overload
+    @staticmethod
+    def analyze_type(__type: Any, *, config: CoreConfig | None = None, parent_depth: int = 2) -> AnalyzedType[Any]:
+        ...
+
+    @staticmethod
+    def analyze_type(__type: Any, *, config: CoreConfig | None = None, parent_depth: int = 2) -> AnalyzedType[Any]:
+        """Generate a pydantic-core schema and other metadata
+        that can be fed into other functions or Validator.
+
+        >>> from typing import List
+        >>> from pydantic import AnalyzedType
+        >>> schema = AnalyzedType.analyze_type(List[int])
+        >>> assert validate(schema, ['123']) == [123]
+
+        Args:
+            __type: The type to generate a core schema for.
+            config: The pydantic-core config. Defaults to None.
+
+        Returns:
+            core_schema.CoreSchema: A pydantic-core CoreSchema.
+        """
+        # if this is a model with an existing schema short circuit
+        # this works for BaseModel and dataclasses
+        if hasattr(__type, '__pydantic_core_schema__') and hasattr(__type, 'model_config'):
+            return AnalyzedType[Any](getattr(__type, '__pydantic_core_schema__'), getattr(__type, 'model_config'))
+        merged_config: CoreConfig = {
+            **(config or {}),  # type: ignore[misc]
+            **getattr(__type, 'model_config', {}),
+        }
+        arbitrary_types = bool((config or {}).get('arbitrary_types_allowed', False))
+        local_ns = _typing_extra.parent_frame_namespace(parent_depth=parent_depth)
+        # BaseModel uses it's own __module__ to find out where it was defined
+        # and then look for symbols to resolve forward references in those globals
+        # On the other hand this function can be called with arbitrary objects,
+        # including type aliases where __module__ (always `typing.py`) is not useful
+        # So instead we look at the globals in our parent stack frame
+        # This works for the case where this function is called in a module that
+        # has the target of forward references in its scope but
+        # does not work for more complex cases
+        # for example, take the following:
+        #
+        # a.py
+        # ```python
+        # from typing import List, Dict
+        # IntList = List[int]
+        # OuterDict = Dict[str, 'IntList']
+        # ```
+        #
+        # b.py
+        # ```python
+        # from pydantic import AnalyzedType
+        # from a import OuterDict
+        # IntList = int  # replaces the symbol the forward reference is looking for
+        # v = AnalyzedType.analyze_type(OuterDict)
+        # v({"x": 1})  # should fail but doesn't
+        # ```
+        #
+        # If OuterDict were a BaseModel this would work because it would resolve
+        # the forward reference within the `a.py` namespace.
+        # But `AnalyzedType.analyze_type(OuterDict)`
```

Comment:
Just reflecting on this a bit, maybe we really should be accepting a namespace argument here _and_ in model_rebuild (rather than a parent namespace depth). I think the parent namespace thing is a good default behavior in most cases but accepting an arbitrary set of definitions does seem like it kind of resolves this in whatever edge case you can come up with.

Maybe a better API (here and elsewhere) would be for this argument to have type `int | dict[str, Any]`, and if you pass a dict it doesn't do any sys.getframe stuff and just uses the passed dict directly.

When I first started thinking about all this namespace stuff I understood it a lot less than I do now. (Not that I would say I understand it fully now..)

Doesn't need to be fixed in this PR, but I think we should maybe revisit this before the official v2 release.